Add vectorized Jacobian and Hessian computation with forward AD

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68629
* #68628
* #67045
* __->__ #67041
* #67820
* #66294
* #66293
* #66292

Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

soulitzer committed:
Add vectorized Jacobian and Hessian computation with forward AD

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -6781,10 +6781,15 @@ def test_jacobian_create_graph(self):
+
+        if test_forward_ad:
+            result_forward_mode = autogradF.jacobian(f, inputs, forward_ad=True, vectorize=True)
+            print(result_forward_mode, expected)
```
debug print

albanD commented on test/test_autograd.py:
```diff
@@ -6856,13 +6861,18 @@ def f(x, y):
         x = torch.randn(3)
         y = torch.randn(3)
-        self._check_jacobian_vectorize_correctness(f, (x, y))
+        # The Jacobin computed using forward AD has the dtype of the output
```
Is that a problem? Should we document that in the jacobian function?

soulitzer replied:
Not sure if it is a problem. But yeah, it's probably good to document.

albanD commented on torch/autograd/functional.py:
```diff
@@ -408,7 +421,66 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
+                                          "jacobian")
+    if vectorize:
+        # See NOTE: [Computing jacobian with vmap and grad for multiple outputs]
+        if strict:
```
This is already checked above.

albanD commented on torch/autograd/functional.py:
```diff
@@ -408,7 +421,66 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
+                           'Please either set `strict=False` or '
+                           '`forward_ad=False`.')
+    is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "jacobian")
+    outputs = func(*inputs)
```
Why do you need this? Your jvp function below can also return the output no?

soulitzer replied:
There needed to be some more logic to extract the output from the wrapped vmap(dual, but it is fixed now

albanD commented on torch/autograd/functional.py:
```diff
@@ -472,6 +544,13 @@ def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
+        if create_graph:
+            raise RuntimeError('torch.autograd.functional.jacobian: `create_graph=True` '
+                               'and `forward_ad=True` are not supported together (yet). '
+                               'Please either set `strict=False` or '
```
That's not the right recommendation?

albanD commented on torch/autograd/functional.py:
```diff
@@ -472,6 +544,13 @@ def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
     """
+    if forward_ad:
+        if create_graph:
+            raise RuntimeError('torch.autograd.functional.jacobian: `create_graph=True` '
```
Why is this one not supported? Just because we don't test it?
Also should this be a NotImplementedError ?

soulitzer replied:
Yeah I haven't tested it. Just putting NotImplementedError for now

albanD commented on torch/autograd/functional.py:
```diff
@@ -690,11 +769,14 @@ def ensure_single_output_function(*inp):
 
     def jac_func(*inp):
+        if forward_ad:
+            # Ensure the graph is created (and not detached)
```
What does this mean? Is that for the create_graph case?

soulitzer replied:
Currently for  in this wrapper make sure inputs have requires_grad=True, and still create_graph=True so that input's won't be detached in jacobian's preprocessing (which detaches when either create_graph=False or requires_grad of its input is False). This way avoids changing jacobian's API but has more overhead because we must create the graph.

Alternatively we could add a for_forward_ad_hessian flag which will skip the _grad_preprocess stage and avoid detaching the inputs while also avoiding the creation of the backward graph

albanD replied:
Ho this was done before the idea that detach applies both to forward and backward mode AD

Should we replace the `.detach()` by a 
```
with torch.no_grad():
    inp = inp.clone()
```
when forward AD is used?
I guess not and we should introduce a new API to detach backward grads only <- TODO for forward AD project

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

albanD reviewed (approved):
some lint failures but lgtm otherwise

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -598,7 +679,7 @@ def vjp(grad_output):
         return _tuple_postprocess(jacobian, (is_outputs_tuple, is_inputs_tuple))
 
-def hessian(func, inputs, create_graph=False, strict=False, vectorize=False):
+def hessian(func, inputs, create_graph=False, strict=False, vectorize=False, use_forward_ad=False):
```
You probably want to name this something more descriptive. If I see `hessian(..., use_forward_ad=True)`, then I might assume that we're using forward-mode AD for both jacobian computations. Maybe "use_forward_ad_for_outer_jacobian"? Or "outer_jacobian_strategy = {forward, reverse}"

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -408,7 +421,64 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
+                                  "only implemented for `vectorize=True`.")
+
+
+def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False, use_forward_ad=False):
```
`use_forward_ad` should probably be more neutral, like "jacobian(strategy={"forward-mode", "reverse-mode"})"

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

zou3519 commented:
By the way, I'm not sure that "strategy" is the right kwarg (it was the first thing that came to mind), should we try to bikeshed it some more?

soulitzer commented:
> By the way, I'm not sure that "strategy" is the right kwarg (it was the first thing that came to mind), should we try to bikeshed it some more?

I guess more generic alternatives could be: `"mode"`, `"method"`, or `"algorithm"`. Doing a quick search of how we name str fields `(str):` or `(str, optional):`, I see that `Upsample`, and `qr`  uses `"mode"` and this actually makes me think we shouldn't use `"mode"` because in both of those cases "mode" indicates *what* is computed as opposed to *how* it is computed. For that reason, I think  `"method"` (which lobpcg uses) or `"algorithm"`/`"strategy"` could be a better choice. I'm sure @albanD has some ideas on this too when he comes back.

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -408,7 +421,64 @@ def _construct_standard_basis_for(tensors: Tuple[torch.Tensor, ...], tensor_nume
+            for jac, input_j in zip(jac.split(input_numels, dim=0), inputs):
+                # We need to transpose the Jacobian because in forward AD, the
+                # batch dimension represents that of the inputs
+                jacobian_input_i_output_j = jac.permute(*range(1, jac.ndim), 0) \
+                    .reshape(tuple([*output_i.shape, *input_j.shape]))  # noqa: C409
```
This is just `jac.movedim(0, -1).unflatten(-1, *input_j.shape)`

zou3519 reviewed (approved):
code looks correct; might want to wait on Alban for bikeshedding

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -6607,7 +6616,7 @@ def f(x, y):
         # output unrelated to all inputs
         def f(x, y):
-            return torch.randn([])
+            return torch.ones([])
```
Why this change?
Or is it just that randomness is never working with vmap? Would a `randn_like(x)` be better?

soulitzer replied:
Yep, it is because randomness does not work with vmap

albanD commented on torch/autograd/functional.py:
```diff
@@ -1,11 +1,24 @@
 import torch
 from typing import Tuple, List
+from . import forward_ad as fwAD
+from torch._vmap_internals import _vmap  # can we replace with use of is_grads_batched=True
```
Not sure wht the comment means?

soulitzer replied:
Good catch, not sure why that is still here. It was from an earlier PR in the stack. I'm removing it.

albanD commented on torch/autograd/functional.py:
```diff
@@ -472,6 +547,14 @@ def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
          tensor([[3., 0.],
                  [0., 3.]]))
     """
+    assert strategy in ("forward-mode", "reverse-mode")
```
There should be a better error message here

albanD commented on torch/autograd/functional.py:
```diff
@@ -675,6 +765,7 @@ def hessian(func, inputs, create_graph=False, strict=False, vectorize=False):
     """
 
     is_inputs_tuple, inputs = _as_tuple(inputs, "inputs", "hessian")
+    assert outer_jacobian_strategy in ("forward-mode", "reverse-mode")
```
same error message

albanD commented on torch/autograd/functional.py:
```diff
@@ -433,6 +503,10 @@ def jacobian(func, inputs, create_graph=False, strict=False, vectorize=False):
             because this feature is still experimental, there may be performance
             cliffs. See :func:`torch.autograd.grad`'s ``batched_grad`` parameter for
             more information.
+        strategy (str, optional): Set to ``"forward-mode"`` or ``"reverse-mode"`` to
```
I think I am happy with both strategy or mode.

One question though is if forward/reverse the right toggle here. Isn't a toggle that talks about the number of input/outputs of the function easier to understand by the end user?

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).

soulitzer committed:
Update on "Add vectorized Jacobian and Hessian computation with forward AD"


Original PR here: https://github.com/pytorch/pytorch/pull/62246 (The old PR does more things, but now that's split across this stack)

This PR:
- Adds "jacfwd" and "hessian_fwdrev"
- Modifies existing tests to also test the `forward_ad=True` case

Differential Revision: [D32314424](https://our.internmc.facebook.com/intern/diff/D32314424)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32314424).
